### PR TITLE
Invoke foundation tests via SwiftPM

### PR DESCRIPTION
**Pull request by jmschonfeld**
This adds new phases to the build script that will build and run swift-foundation unit tests and swift-corelibs-foundation unit tests via swiftpm. On platforms that use the `build-script`, this will take place later than before after the SwiftPM build has finished. On platforms that use `build.ps1` the tests will happen as they did previously but will build with SwiftPM instead. We set the `SWIFTCI_USE_LOCAL_DEPS` environment variable when building these projects so that they find dependencies in the local checkout rather than fetching them from GitHub.

### CODE REVIEW

**Comment by jmschonfeld** [comment]
@swift-ci please smoke test

**Comment by jmschonfeld** [comment]
@swift-ci please smoke test

**Comment by jmschonfeld** [comment]
@swift-ci please smoke test

**Comment by jmschonfeld** [comment]
https://github.com/apple/swift-foundation/pull/756

@swift-ci please smoke test

**Comment by jmschonfeld** [comment]
https://github.com/apple/swift-foundation/pull/761

@swift-ci please smoke test

**Comment by jmschonfeld** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/5025

@swift-ci please smoke test

**Comment by jmschonfeld** [comment]
https://github.com/apple/swift-corelibs-foundation/pull/5025

@swift-ci please smoke test macOS platform

**Comment by jmschonfeld** [comment]
This is now ready to go / ready for review! Just a note: this does not yet enable these tests on Windows yet, I'm still working through a few issues with getting those building correctly (the tests can be invoked but do not pass yet and so I've left them disabled in CI for now, but would like to add this ability so that it can be tested and iterated on locally with ease). In the linux CI run, you can see the tests for swift-foundation and swift-corelibs-foundation running and passing.

File: utils/build.ps1 [hunk 1]
```diff
@@ -1618,26 +1618,41 @@ function Build-Dispatch([Platform]$Platform, $Arch, [switch]$Test = $false) {
...
+    Isolate-EnvVars {
+      $env:SWIFTCI_USE_LOCAL_DEPS=1
+      Build-SPMProject `
+        -Test `
+        -Src $SourceCache\swift-foundation `
+        -Bin $OutDir `
+        -Arch $HostArch
+    }
+
+    $OutDir = Join-Path -Path $HostArch.BinaryCache -ChildPath foundation-tests
```
Comment (compnerd) [inline utils/build.ps1:1634]: Why recompute this?

File: utils/build.ps1 [hunk 2]
```diff
@@ -1618,26 +1618,41 @@ function Build-Dispatch([Platform]$Platform, $Arch, [switch]$Test = $false) {
...
+    $ShortArch = $Arch.LLVMName
+
+    Isolate-EnvVars {
       $TestingDefines = @{ ENABLE_TESTING = "NO" }
       $Targets = @("default", "install")
       $InstallPath = "$($Arch.SDKInstallRoot)\usr"
-    }
+
 
     $env:CTEST_OUTPUT_ON_FAILURE = 1
```
Comment (compnerd) [inline utils/build.ps1:1657]: This variable is no longer useful if we are no longer using CMake for the tests.

File: utils/build.ps1 (hunk 1 above)
Comment (jmschonfeld) [reply utils/build.ps1:1634]: This is different than the first `$OutDir` - the former is `/swift-foundation-tests` and the latter is `/foundation-tests` to use separate binary directories for the two projects

**Comment by jmschonfeld** [comment]
@swift-ci please smoke test

**Comment by parkera** [comment]
In that case, there would be some indication at test time that swiftpm failed to run the tests - a crash, for example, right?

**Comment by finagolfin** [comment]
Many different possibilities, but there's a reason this `swift-corelibs-foundation` test run on the CI was careful to run the tests before SwiftPM, even though it had to build XCTest first then come back to run the `swift-corelibs-foundation` tests.

I wonder if we can keep that by leveraging the prebuilt release SwiftPM on the CI to invoke the freshly-built trunk Swift compiler and keep running the `swift-corelibs-foundation` tests right after building them, that's all.

**Comment by jmschonfeld** [comment]
> have you considered using the prebuilt SwiftPM with the freshly-built Swift compiler to build and run these tests right away?

I don't think this is a viable option (currently) for swift-foundation since the CI images have the 5.8 toolchain AFAIK and swift-foundation relies on Swift macros which are only supported in the SwiftPM tools packaged with swift 5.9+.

I see where you're coming from here but I'm also not sure it'll be too large of an issue in practice. We expect that at some point soon once we re-configure our CI, changes to swift-foundation/swift-corelibs-foundation will be able to run SwiftPM tests in CI directly using a nightly toolchain without requiring a full toolchain build (in addition to the option to run full toolchain tests). So ideally if something here breaks in a full toolchain test, we'd have the option to run the unit tests against the nightly swift toolchain to diagnose further in CI.

This change as it stands at least gets us closer to where we want to be re. testing Foundation instead of avoiding testing at all like we do today, and if in the future we have a pre-built toolchain of a compatible version and mixing-and-matching was supported fully on the SwiftPM side then we could investigate that as a future incremental followup, but IMO this is a nice first step that we can take.

**Comment by finagolfin** [comment]
Sounds good, good point about SwiftPM 5.8.1 on the CI not supporting macros yet.

**Comment by finagolfin** [comment]
@compnerd, if you would review, we can get these tests running on the CI again.

**Review by compnerd — APPROVED** [review]

**Review by iCharlesHu — APPROVED** [review]

**Pull request merged**